agents/maintenance_scheduler/maintenance_scheduler/tools/tools.py

```diff
@@ -0,0 +1,279 @@
+
+      Returns:
+          A dictionary, where the key is the bus stop id  and the value is the list
+          of expected number of passengers at a particular point in time.
+
+      Example:
+          >>> get_expected_number_of_passengers(bus_stop_ids=['bus-stop-1', 'bus-stop-2'])
+          {"bus-stop-1":
+              [{"time": "2025-04-21 18:02:33.463897+00:00", "number_of_passengers":	13},
+              {"time": "2025-04-21 18:07:33.463897+00:00", "number_of_passengers":	15},
+              {"time": "2025-04-21 18:08:33.463897+00:00", "number_of_passengers":	4}],
+          "bus-stop-2":
+              [{"time": "2025-04-21 18:02:33.463897+00:00", "number_of_passengers":	5},
+              {"time": "2025-04-21 18:07:33.463897+00:00", "number_of_passengers":	7},
+              {"time": "2025-04-21 18:08:33.463897+00:00", "number_of_passengers":	10}]
+          }
+      """
+    logger.info("Retrieving expected number of passengers for %s", bus_stop_ids)
+
+    if config.mock_tools:
+        result = {}
+        for bus_stop_id in bus_stop_ids:
+            forecast = []
+            base_number_of_passengers = random.randint(5, 20)
+            for next_increment in range(10, 3 * 24 * 60, 15):
+                forecast.append({'time': (
+                    datetime.now(tz=time_zone) + timedelta(
+                    minutes=next_increment)
+                ).isoformat(), 'number_of_passengers': (
+                    base_number_of_passengers + random.randint(3, 10))})
+            result[bus_stop_id] = forecast
+        return result
+
+    job_config = bigquery.QueryJobConfig(
+        query_parameters=[
+            bigquery.ArrayQueryParameter('bus_stop_ids', "STRING", bus_stop_ids)
+        ]
+    )
+
+    rows = bigquery_client.query_and_wait(
+        job_config=job_config,
+        project=config.get_bigquery_run_project(),
+        query=f"""
+        WITH forecast AS (
+            SELECT
+              bus_stop_id, forecast_timestamp, 
+              CAST(forecast_value AS INT64) as expected_number_of_passengers
+            FROM
+              AI.FORECAST(
+                (SELECT bus_stop_id, event_ts, num_riders
+                  FROM `{config.get_bigquery_data_project()}.bus_stop_image_processing.bus_ridership`
+                  WHERE bus_stop_id IN UNNEST(@bus_stop_ids)),
+                data_col => 'num_riders',
+                timestamp_col => 'event_ts',
+                model => 'TimesFM 2.0',
+                id_cols => ['bus_stop_id'],
+                horizon => 500,
+                confidence_level => .8)
+        )
+        SELECT bus_stop_id, forecast_timestamp, expected_number_of_passengers 
+            FROM forecast WHERE forecast_timestamp > CURRENT_TIMESTAMP()"""
+    )
+
+    result = {}
+    for row in rows:
+        bus_stop_id = row.bus_stop_id
+        forecast = result[bus_stop_id] if bus_stop_id in result else []
+        forecast.append(
+            {'time': row.forecast_timestamp
```

Comment:
would like to see the dot chain wrapped in parens here.

Reply:
I didn't touch this one - seems like it's working fine and adding more parens will make it less readable. Feel free to suggest.